### PR TITLE
fix: handle empty candidate sets in ISMAGS.largest_common_subgraph

### LINKED ISSUE #8885: ISMAGS raises ValueError if no nodes are compatible
### Current Behavior

Hi!
If I do something silly, and ask ismags for the largest common subgraph between a graph and subgraph that have no compatible nodes due to attribute mismatches, I get presented with a fairly gnarly ValueError:

```py
graph = nx.Graph()
graph.add_nodes_from([0])

subgraph = nx.Graph()
subgraph.add_nodes_from({0: {'attr1': 0}}.items())

node_match = nx.isomorphism.categorical_node_match(['attr1'], [None])

ismags = nx.algorithms.isomorphism.ISMAGS(graph, subgraph, node_match=node_match)
symmetric = list(ismags.largest_common_subgraph(False))
```
Gives:
```
Traceback (most recent call last):
  File "/homes/peterkroon/Documents/vermouth-martinize/derp/ismags_mess.py", line 42, in <module>
    symmetric = list(ismags.largest_common_subgraph(False))
  File "/homes/peterkroon/Documents/networkx/networkx/algorithms/isomorphism/ismags.py", line 842, in largest_common_subgraph
    yield from self._largest_common_subgraph(
        candidate_sets, constraints, to_be_mapped
    )
  File "/homes/peterkroon/Documents/networkx/networkx/algorithms/isomorphism/ismags.py", line 1271, in _largest_common_subgraph
    next_sgn = min(nodes, key=lambda n: min(len(x) for x in candidates[n]))
ValueError: min() iterable argument is empty

```

### Expected Behavior
I would prefer an empty dict as answer

### Environment

NetworkX version: master branch (8e0b38e1a11335f07db13a13a6a5955eecf33408)

### PR DESCRIPTION
**Pull request by mturac**
fixes #8885

`largest_common_subgraph()` crashes with `ValueError: min() iterable argument is empty` when subgraph and graph dont share any node colors. the guard on line 839 was checking `any(candidate_sets.values())` wich is allways truthy because the values are singleton sets even when the inner frozenset is empty.

fixed it to check `any(to_be_mapped)` insted — if theres nothing to map, yield nothing and return. added two tests for the case where no colors overlap.

### CODE REVIEW

File: networkx/algorithms/isomorphism/tests/test_ismags.py [hunk 1]
```diff
@@ -634,6 +634,41 @@ def test_extra_colors(self):
         assert ismags._sgn_partition == [{1}, {0}, {2}]
         assert ismags._gn_partition == [{5}, set(), set(), {2}]
 
+    def test_no_common_colors(self):
```
Comment (dschult) [inline networkx/algorithms/isomorphism/tests/test_ismags.py:637]: Let's test for colors and self-loops. Both conditions will result in nodes with no candidates.

Something like:
```python
    def test_color_mismatches(self):
        # see gh-8885
        graph = nx.path_graph(5)
        subgraph = nx.path_graph(5)
        for n in subgraph:
            subgraph.nodes[n]["color"] = "blue"

        # check that normal case works as expected
        ismags = iso.ISMAGS(graph, subgraph)
        assert list(ismags.largest_common_subgraph()) == [{i: i for i in subgraph}]

        # check case when a subgraph node has no candidate color match
        nodematch = nx.isomorphism.categorical_node_match("color", None)
        ismags = iso.ISMAGS(graph, subgraph, node_match=nodematch)
        assert list(ismags.largest_common_subgraph()) == []

        # check case when a subgraph node has no candidate sue to selfloops
        for n in graph:
            graph.add_edge(n, n)
        ismags = iso.ISMAGS(graph, subgraph)
        assert list(ismags.largest_common_subgraph()) == []
 ```

File: networkx/algorithms/isomorphism/ismags.py [hunk 2]
```diff
@@ -836,13 +836,16 @@ def largest_common_subgraph(self, symmetry=True):
 
         candidate_sets = self._get_node_color_candidate_sets(MONO_fits=operator.eq)
 
-        if any(candidate_sets.values()):
-            relevant_parts = self._sgn_partition[: self.N_node_colors]
-            to_be_mapped = {frozenset(n for p in relevant_parts for n in p)}
+        relevant_parts = self._sgn_partition[: self.N_node_colors]
+        to_be_mapped = {frozenset(n for p in relevant_parts for n in p)}
+
+        if any(to_be_mapped):
```
Comment (dschult) [inline networkx/algorithms/isomorphism/ismags.py:842]: This no longer checks whether any node has a candidate.  
They might not have a candidate for node color or for number of self-loops.
What do you think of the following (leaving other parts unchanged--that is indented):  
```python
        if any(all(constraint) for constraint in candidate_sets.values()):
 ```

**Review by dschult — COMMENTED** [review]
Thanks for this!  Unfortunately this doesn't cover all the cases that the old code covered. The nodes can have no candidates due to either node color or number of self-loops.  suggestions below.

**Review by dschult — APPROVED** [review]
I took the liberty of adding a few corner case tests for largest common subgraph:
- null graphs
- small empty graphs (that either do or don't have color mismatches)
- I changed the PR test of multi-node but no color matches to test both color matches and self-loop matches.

For the fix itself, I updated the if-clause to check for self-loop and/or color no-matches and I tightened up comments somewhat.

 I approve this PR.   Thanks @mturac
